sqldelight-compiler/src/test/kotlin/app/cash/sqldelight/core/queries/async/AsyncSelectQueryTypeTest.kt

```diff
@@ -260,4 +260,218 @@ class AsyncSelectQueryTypeTest {
       """.trimMargin(),
     )
   }
+
+  @Test
+  fun `IN in async postgresql queries uses numbered indices`() {
+    val result = FixtureCompiler.parseSql(
+      """
+      |CREATE TABLE data (
+      |  id INTEGER PRIMARY KEY,
+      |  value TEXT
+      |);
+      |
+      |selectForMultipleIds:
+      |SELECT *
+      |FROM data
+      |WHERE id IN ?;
+      """.trimMargin(),
+      tempFolder,
+      dialect = PostgreSqlDialect(),
+      generateAsync = true,
+    )
+
+    val query = result.namedQueries.first()
+    val generator = SelectQueryGenerator(query)
+
+    assertThat(generator.querySubtype().toString()).isEqualTo(
+      """
+        |private inner class SelectForMultipleIdsQuery<out T : kotlin.Any>(
+        |  public val id: kotlin.collections.Collection<kotlin.Int>,
+        |  mapper: (app.cash.sqldelight.db.SqlCursor) -> T,
+        |) : app.cash.sqldelight.Query<T>(mapper) {
+        |  public override fun addListener(listener: app.cash.sqldelight.Query.Listener): kotlin.Unit {
+        |    driver.addListener(listener, arrayOf("data"))
+        |  }
+        |
+        |  public override fun removeListener(listener: app.cash.sqldelight.Query.Listener): kotlin.Unit {
+        |    driver.removeListener(listener, arrayOf("data"))
+        |  }
+        |
+        |  public override fun <R> execute(mapper: (app.cash.sqldelight.db.SqlCursor) -> R): app.cash.sqldelight.db.QueryResult<R> {
+        |    val idIndexes0 = createNumberedArguments(offset = 0, count = id.size)
+        |    return driver.executeQuery(null, ""${'"'}
+        |        |SELECT *
+        |        |FROM data
+        |        |WHERE id IN ${"$"}idIndexes0
+        |        ""${'"'}.trimMargin(), mapper, id.size) {
+        |          check(this is app.cash.sqldelight.driver.r2dbc.R2dbcPreparedStatement)
+        |          id.forEachIndexed { index, id_ ->
+        |            bindLong(index, id_.toLong())
+        |          }
+        |        }
+        |  }
+        |
+        |  public override fun toString(): kotlin.String = "Test.sq:selectForMultipleIds"
+        |}
+        |
+      """.trimMargin(),
+    )
+  }
+
+  @Test
+  fun `offset calculation for IN in async postgresql queries is fine`() {
+    val result = FixtureCompiler.parseSql(
+      """
+    |CREATE TABLE data (
+    |  id INTEGER PRIMARY KEY,
+    |  value TEXT
+    |);
+    |
+    |selectForValueAndMultipleIds:
+    |SELECT *
+    |FROM data
+    |WHERE value = ? AND id IN ?;
```

Comment:
Could you add another `?` after the array parameter just to make sure the index still works?

Reply:
Good catch. It actually doesn't work as intended. It changes the `?` behind the list array $2 but tries to bind at index `id.size+1`. I think the best way to fix this is by moving the non array arguments to the front and change the way the offset calculation and the code for the binding works accordingly.